src/xtgeo/interfaces/rms/_regular_surface.py

```diff
@@ -0,0 +1,449 @@
+
+        # Do the write
+        self._write_to_rms(data, validated_values)
+
+        # Save project if external
+        assert rmsapi is not None
+        proj = self._rmsapi_utils.project
+        assert isinstance(proj, rmsapi.Project), "Project must be initialized"
+        if getattr(self._rmsapi_utils, "_rmsexternal", False):
+            proj.save()
+
+    def _check_valid_domain(self) -> None:
+        """Check that domain is valid for the given storage type."""
+        if self.domain not in DomainType.values():
+            raise ValueError(f"domain must be {DomainType.values()}")
+        self._domain_enum = DomainType[self.domain.upper()]
+        self._api_vertical_domain = self._resolve_api_vertical_domain()
+        logger.debug("Using domain: %s", self._domain_enum)
+
+    def _resolve_api_vertical_domain(self) -> Any:  # rmsapi.VerticalDomain
+        """Map internal domain enum to RMS API VerticalDomain enum."""
+        assert rmsapi is not None
+        if self._domain_enum == DomainType.DEPTH:
+            return rmsapi.VerticalDomain.depth
+        if self._domain_enum == DomainType.TIME:
+            return rmsapi.VerticalDomain.time
+        return rmsapi.VerticalDomain.unknown
+
+    @staticmethod
+    def _validate_payload(data: RegularSurfaceDataRms) -> np.ma.MaskedArray:
+        """Validate payload and return a normalized copy of values.
+
+        This method validates the surface data and returns a copy of the values
+        array normalized to MaskedArray format, without mutating the input.
+
+        Args:
+            data: The surface data to validate
+
+        Returns:
+            A copy of the values array as np.ma.MaskedArray
+
+        Raises:
+            ValueError: If validation fails
+        """
+        if data.ncol <= 0 or data.nrow <= 0:
+            raise ValueError("ncol and nrow must be positive.")
+        if data.xinc <= 0 or data.yinc <= 0:
+            raise ValueError("xinc and yinc must be positive.")
+
+        # Normalize to MaskedArray without mutating input
+        if not isinstance(data.values, np.ma.MaskedArray):
+            values = np.ma.array(data.values, copy=True)
+        else:
+            # Even if already MaskedArray, we'll make a copy to be safe
+            values = data.values.copy()
+
+        if values.shape != (data.ncol, data.nrow):
+            raise ValueError(
+                f"values shape {values.shape} does not match "
+                f"(ncol, nrow)=({data.ncol}, {data.nrow})"
+            )
+
+        return values
+
+    def _write_to_rms(
+        self, data: RegularSurfaceDataRms, values: np.ma.MaskedArray
+    ) -> None:
+        assert rmsapi is not None
+        assert self._rmsapi_utils is not None, "RMS API utils must be initialized"
+        proj = self._rmsapi_utils.project
+        assert isinstance(proj, rmsapi.Project), "Project must be initialized"
+        assert self._stype_enum is not None, "stype must be set"
```

Comment:
These kinds of asserts shouldn't be necessary for properties in most cases. It's of course valid to have unintialized properties in some circumstances, but they should either be initialized in the `init` method if possible, or lazily initialized when wrapped by a property or method getter.

If the type checker cannot know the state of something, that essentially means the programmer will also have a hard time knowing it and has to manually hold the state machine in their head when working with this class/base class.